File: packages/ui/src/memberships.test.ts
import { describe, expect, test } from "bun:test";
import type { MembershipDto } from "@society-hub/types";
import { uniqueMembershipsBySociety } from "./memberships";

describe("uniqueMembershipsBySociety", () => {
  test("keeps one Keshav Heights row and prefers chairperson", () => {
    const rows: MembershipDto[] = [
      {
        tenantId: "t1",
        societyName: "Keshav Heights",
        role: "committee",
        canUseAdminMode: true,
      },
      {
        tenantId: "t1",
        societyName: "Keshav Heights",
        role: "chairperson",
        canUseAdminMode: true,
      },
      {
        tenantId: "t2",
        societyName: "Other Society",
        role: "resident",
        canUseAdminMode: false,
      },
    ];
    expect(uniqueMembershipsBySociety(rows)).toEqual([
      {
        tenantId: "t1",
        societyName: "Keshav Heights",
        role: "chairperson",
        canUseAdminMode: true,
      },
      {
        tenantId: "t2",
        societyName: "Other Society",
        role: "resident",
        canUseAdminMode: false,
      },
    ]);
  });

  test("skips the picker when only one society has two roles", () => {
    const rows: MembershipDto[] = [
      {
        tenantId: "t1",
        societyName: "Keshav Heights",
        role: "committee",
        canUseAdminMode: true,
      },
      {
        tenantId: "t1",
        societyName: "Keshav Heights",
        role: "chairperson",
        canUseAdminMode: true,
      },
    ];
    expect(uniqueMembershipsBySociety(rows)).toHaveLength(1);
  });
});

File: apps/mobile/test/unit/memberships_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:societyhub_mobile/api/models.dart';
import 'package:societyhub_mobile/auth/memberships.dart';

MembershipDto _m({
  required String tenantId,
  required String societyName,
  required String role,
}) {
  return MembershipDto(
    tenantId: tenantId,
    societyName: societyName,
    role: role,
    canUseAdminMode: role != 'resident',
  );
}

void main() {
  test('keeps one option per society and prefers chairperson', () {
    final unique = uniqueMembershipsBySociety([
      _m(tenantId: 't1', societyName: 'Keshav Heights', role: 'committee'),
      _m(tenantId: 't1', societyName: 'Keshav Heights', role: 'chairperson'),
      _m(tenantId: 't2', societyName: 'Other Society', role: 'resident'),
    ]);
    expect(unique, hasLength(2));
    expect(unique.first.tenantId, 't1');
    expect(unique.first.role, 'chairperson');
    expect(unique.last.tenantId, 't2');
  });
}
